### PR TITLE
Graph Changes

### PR DESCRIPTION
**Pull request by Havvy**
General cleanup and adding a few methods that I want to use in Clippy.

Need somebody to bikeshed names.

### CODE REVIEW

File: src/librustc_data_structures/graph/tests.rs [hunk 1]
```diff
@@ -139,3 +165,15 @@ fn each_adjacent_from_d() {
...
+fn is_node_cyclic_a() {
+    let graph = create_graph_with_cycle();
+    assert!(!graph.is_node_cyclic(NodeIndex(0)));
+}
+
+#[test]
+fn is_node_cyclic_b() {
+    let graph = create_graph_with_cycle();
+    assert!(graph.is_node_cyclic(NodeIndex(1)));
+}
```
Comment (eddyb) [inline src/librustc_data_structures/graph/tests.rs]: Missing newline at the end of the file.

**Comment by eddyb** [comment]
cc @nikomatsakis Are we fine with such additions.

File: src/librustc_data_structures/graph/tests.rs (hunk 1 above)
Comment (Havvy) [inline src/librustc_data_structures/graph/tests.rs]: Added newline.

**Comment by bluss** [comment]
here it comes: Using the convention of naming the iterator struct exactly like the method that creates it would be good. I see that the existing outgoing_edges doesn't do that, but it's not too late for new additions to do so.

File: src/librustc_data_structures/graph/tests.rs [hunk 2]
```diff
@@ -20,10 +20,13 @@ fn create_graph() -> TestGraph {
 
     // Create a simple graph
     //
+    //          F
+    //          |
+    //          V
     //    A -+> B --> C
```
Comment (ranma42) [inline src/librustc_data_structures/graph/tests.rs]: This should probably become
```
     //    A --> B --> C
```
(i.e. '+' should become '-')

**Comment by Havvy** [comment]
@bluss So based on that, what names are you thinking? Because I think you're thinking `nodes_iter_enumerated` or `nodes_iter_enumerated_by_index`.

**Comment by bluss** [comment]
@Havvy Well I didn't propose anything about that, only that the struct should be named like the method or vice versa. So for example `AllNodesEnumerated` with the current patch.

Spontaneously, maybe `fn enumerated_nodes(&self) -> EnumeratedNodes<N>` and so on?

**Comment by Havvy** [comment]
I like those method names better, and went with them, and also made the Iterator structs match the method names.

**Comment by nikomatsakis** [comment]
I have no *objection* to extending the graph API, but I would say we have no intention of supporting it. It'd probably be better for clippy to use petgraph or something on `crates.io`.

**Comment by eddyb** [comment]
@nikomatsakis This is for using `rustc::cfg` directly. Maybe I should've suggested MIR, but #37400 hadn't landed yet at that time.

**Comment by nikomatsakis** [comment]
@eddyb it'd probably be better for clippy to recreate the graph in petgraph then =)

**Comment by nikomatsakis** [comment]
but yeah I'm fine w/ landing these changes, particularly since `cfg`'s time is limited.

**Comment by Havvy** [comment]
Is `rustc_data_structures::graph` also going away when `cfg` goes away? If not, these are still probably useful.

**Comment by alexcrichton** [comment]
@Havvy thoughts about extracting this data structure to crates.io, perhaps build on petgraph? If what @nikomatsakis [says](https://github.com/rust-lang/rust/pull/37535#issuecomment-258497197) is correct then this is soon to be deleted anyway.

**Comment by Havvy** [comment]
There's a similar lint already in the compiler (infinitely recursive functions) as to what I am attempting to do (though I've put in on hold for a couple days because, yay trying to figure out what's staying and what's changing), so I can just use what that lint uses. I don't need `cfg` if there's an alternative way and I'd rather not maintain or force to be maintained for clippy a version of `cfg` outside of the Rust repository.

So my question is, is the Graph structure used outside of cfg. If so, these commits are still probably useful (except probably the last one), but if not, there's no point merging this and it should be closed. So, what is the future of Graph?

**Comment by nikomatsakis** [comment]
To be clear, I have no objection to landing this PR for the time being. I just want to be clear that there are no promises of backwards compatibility etc.

**Pull request merged**